### PR TITLE
Losses and metrics improvements (mostly)

### PR DESCRIPTION
**Pull request by ocourtin**
Hi,

Quite a few improvements in this PR:
- Add mIoU loss
- Add Lovasz loss
- Add a config parameter to allow to choose loss function among implemented ones
- Remove useless losses parameters (size_average and ignore_index)
- Add foreground IoU metric
- Add MCC metric
- Refactor and cleanup metric class
- Add resume training option 
- Add log training in text file, including hyperparameters 

As a consequence:
 - Imply to add loss parameter in model config file
 - .pth files already generated are not compliant anymore

Nota:
 - Pixel accuracy metric wasn't that meaningful, i implement it, than remove it.
 - mIoU and Lovasz both significantly improve semseg (rather than x entropy loss) 

Thanks in advance for the code review ^^

### CODE REVIEW

File: robosat/losses.py [hunk 1]
```diff
@@ -10,17 +11,15 @@ class CrossEntropyLoss2d(nn.Module):
     See: http://cs231n.github.io/neural-networks-2/#losses
     """
 
-    def __init__(self, weight=None, size_average=True, ignore_index=255):
+    def __init__(self, weight=None):
```
Comment (daniel-j-h) [inline robosat/losses.py]: I think we had the two parameters here to be in line with what the PyTorch losses were doing. I'm fine with removing them.

File: config/model-unet.toml [hunk 2]
```diff
@@ -29,3 +29,6 @@
 
   # Weight decay l2 penalty for the optimizer
   decay      = 0.0001
+
+  # Loss function name (e.g 'CrossEntropy', 'mIoU', 'Lovasz' or 'Focal')
+  loss = 'CrossEntropy'
```
Comment (daniel-j-h) [inline config/model-unet.toml]: Should we support weighted losses? E.g. a * cross entropy + b * dice? I think that was the initial plan in https://github.com/mapbox/robosat/issues/3. Do you think it makes sense to let user specify that?

I would love if we could have sensible default settings users without domain knowledge in losses should tweak. I'm not sure cross entropy is a good default choice now that we have lovasz for example. Did you run experiments and do you have numbers we can compare?

Also the focal loss should probably not be used until we implement retina net for object detection, and even then we need to benchmark it first.

File: robosat/metrics.py [hunk 3]
```diff
@@ -27,42 +32,40 @@ def add(self, actual, predicted):
...
+        """Retrieves the mean Intersection over Union score.
 
         Returns:
-          The mean intersection over union score for all obersations seen so far.
+          The mean Intersection over Union score for all observations seen so far.
         """
+        return np.nanmean([self.tn / (self.tn + self.fn + self.fp), self.tp / (self.tp + self.fn + self.fp)])
 
-        intersection = np.diag(self.confusion_matrix)
+    def get_fg_iou(self):
```
Comment (daniel-j-h) [inline robosat/metrics.py]: This is great, very good to have!

File: robosat/tools/export.py [hunk 4]
```diff
@@ -32,7 +32,7 @@ def map_location(storage, _):
 
     chkpt = torch.load(args.checkpoint, map_location=map_location)
     net = torch.nn.DataParallel(net)
-    net.load_state_dict(chkpt)
+    net.load_state_dict(chkpt['state_dict'])
```
Comment (daniel-j-h) [inline robosat/tools/export.py]: Would be great if you could format all your changes with black. The travis config shows the command line argument for it; thanks!

File: robosat/tools/train.py [hunk 5]
```diff
@@ -46,11 +46,23 @@ def add_parser(subparser):
...
 
+class Log:
+    def __init__(self, path):
+        self.fp = open(path, "a")
+
+    def log(self, msg):
+        print(msg)
+        assert self.fp, "Unable to write in log file"
+        self.fp.write(msg + os.linesep)
+        self.fp.flush()
```
Comment (daniel-j-h) [inline robosat/tools/train.py]: Probably a good idea to create a new Python module for everything logging related.

If we make it a bit more generic we could also change the impl. e.g. to a `csv.writer` or later on think about hooking up e.g. tensorboard.

File: robosat/tools/train.py [hunk 6]
```diff
@@ -73,39 +85,75 @@ def main(args):
...
+    if model["opt"]["loss"] == "CrossEntropy":
+        criterion = CrossEntropyLoss2d(weight=weight).to(device)
+    elif model["opt"]["loss"] == "mIoU":
+        criterion = mIoULoss2d(weight=weight).to(device)
+    elif model["opt"]["loss"] == "Focal":
+        criterion = FocalLoss2d(weight=weight).to(device)
+    elif model["opt"]["loss"] == "Lovasz":
+        criterion = LovaszLoss2d().to(device)
+    else:
+        sys.exit("Error: Unknown [opt][loss] value !")
```
Comment (daniel-j-h) [inline robosat/tools/train.py]: Hm what's a good way to make this a bit more elegant? Can we e.g. put a `__name__` attribute in the loss classes or use the class name, and then lookup all available losses from a set of classes?

File: config/model-unet.toml (hunk 2 above)
Comment (ocourtin) [reply config/model-unet.toml]: Related to #3

After several attempts (on mIoU)  i finally find that a max(loss1 , loss2) could be an easy way to reach good results with fewer hyperparameters to adjust.
Dice based loss imply to find a good enough tractable approximation, and did'nt take time to look on this at this point.

You're right on the easy to play configuration, and
i just switch from CrossEntropy to Lovasz default loss.

And if we talk about numbers Lovasz and mIoU are close.
They both better handle shape segmentation on near features and reduce probs noise. 
We could even imagine removing one step in the process as probs and masks output are really similars.

X Entropy: mIoU: 0.835, building IoU: 0.735
mIoU:        mIoU: 0.848, building IoU: 0.752 
Lovasz:     mIoU: 0.853, building IoU: 0.759

File: robosat/tools/export.py (hunk 4 above)
Comment (ocourtin) [reply robosat/tools/export.py]: Fixed !

File: robosat/tools/train.py (hunk 5 above)
Comment (ocourtin) [reply robosat/tools/train.py]: Could be yes,
I just pushed a new separated Log module.
At this point, it still quite close the initial implementation.

File: robosat/tools/train.py (hunk 6 above)
Comment (ocourtin) [reply robosat/tools/train.py]: It's quite easy indeed to retrieve all loss available, and pick up the good one if present.
But the point here is that there's no reason that all losses would share the same prototype (i.e lovasz one don't need weights as weighted crossentropy ones).

An idea to deal with that, while keeping the human readibility ?

File: config/model-unet.toml (hunk 2 above)
Comment (ocourtin) [reply config/model-unet.toml]: One compare probs predict sample: ( image / label / x_entropy / mIoU / Lovasz )

File: config/model-unet.toml [hunk 7]
```diff
@@ -29,3 +29,6 @@
 
   # Weight decay l2 penalty for the optimizer
   decay      = 0.0001
+
+  # Loss function name (e.g 'Lovasz', 'mIoU' or 'CrossEntropy')
+  loss = 'Lovasz'
```
Comment (daniel-j-h) [inline config/model-unet.toml]: Thank you, this will be a great default for folks! :bow:

File: robosat/log.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+
+class Log:
+
+    """Create a log instance on a log file
+    """
+
+    def __init__(self, path, stdout=True):
+        self.stdout = stdout
+        self.fp = open(path, "a")
```
Comment (daniel-j-h) [inline robosat/log.py]: Should we `.close` the file descriptor explicitly somewhere?

File: robosat/log.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+import os
+
+
+class Log:
+
+    """Create a log instance on a log file
+    """
+
+    def __init__(self, path, stdout=True):
```
Comment (daniel-j-h) [inline robosat/log.py]: Instead of a boolean what about letting the user pass `sys.stdout` or `sys.stderr` here? Or `None` for don't print.

File: robosat/tools/predict.py [hunk 10]
```diff
@@ -65,7 +65,7 @@ def map_location(storage, _):
     if cuda:
         torch.backends.cudnn.benchmark = True
 
-    net.load_state_dict(chkpt)
+    net.load_state_dict(chkpt['state_dict'])
```
Comment (daniel-j-h) [inline robosat/tools/predict.py]: Looks like here and below we need black formatting.

File: robosat/tools/train.py (hunk 6 above)
Comment (daniel-j-h) [reply robosat/tools/train.py]: Should we create an interface to pass weights to all loss functions? And then write an adapter for lovasz which simply discards the weights and created a lovasz loss?

File: robosat/tools/train.py [hunk 11]
```diff
@@ -73,39 +75,75 @@ def main(args):
...
+        sys.exit("Error: Epoch {} set in {} already reached by the checkpoint provided".format(num_epochs, args.model))
 
     history = collections.defaultdict(list)
+    log = Log(os.path.join(model["common"]["checkpoint"], "log"))
+
+    log.log("--- Hyper Parameters on Dataset: {} ---".format(dataset["common"]["dataset"]))
+    log.log("Batch Size:\t {}".format(model["common"]["batch_size"]))
+    log.log("Image Size:\t {}".format(model["common"]["image_size"]))
+    log.log("Learning Rate:\t {}".format(model["opt"]["lr"]))
+    log.log("LR Decay:\t {}".format(model["opt"]["decay"]))
```
Comment (daniel-j-h) [inline robosat/tools/train.py]: This is not the learning rate decay but instead the l2 regularization decay for Adam; not sure we should use it in the first place since what we really want is AdamW and not the l2 weight decay implemented in PyTorch currently.

Comment (ocourtin) [reply robosat/tools/train.py]: Right now, the only losses prototypes are with or without weights.
But we could imagine adding per batch rather than per image.
Or if Focal is useful adding gamma in userland.
And so on...

So i'm not convinced at this point, that in the end, it would be more elegant/simpler 
to read and maintain...



Thanks for Adam's Weight Decay. And i updated the label accordingly.

File: config/model-unet.toml (hunk 7 above)
Comment (ocourtin) [reply config/model-unet.toml]: :)

File: config/model-unet.toml (hunk 2 above)
Comment (daniel-j-h) [reply config/model-unet.toml]: Fascinating! Amazing work! :tada:

File: robosat/tools/train.py (hunk 6 above)
Comment (daniel-j-h) [reply robosat/tools/train.py]: In addition above we require the user to always calculate and pass in the dataset weights.

Should we still let the user do thiseven if the lovasz loss function doesn't need it?

Otherwise I think this pull request is ready to go - just this small nitpick here.

File: robosat/log.py (hunk 8 above)
Comment (daniel-j-h) [reply robosat/log.py]: Hm, if we flush it should be fine I guess. Was just wondering if we can expose this in a clean way.

File: robosat/tools/train.py (hunk 11 above)
Comment (daniel-j-h) [reply robosat/tools/train.py]: :+1:

File: robosat/tools/predict.py (hunk 10 above)
Comment (daniel-j-h) [reply robosat/tools/predict.py]: No worries; just checked locally and it's fine now!

File: robosat/tools/train.py (hunk 6 above)
Comment (ocourtin) [reply robosat/tools/train.py]: Yes, good point !

Could been indeed improve to allow dataset weights optionnal if choosen loss don't need them...
72a13df commit handle this.

Comment (ocourtin) [reply robosat/tools/train.py]: As everything seems OK now, could you apply the PR ?
Thanks in advance :)

**Comment by daniel-j-h** [comment]
This looks good to go! Thanks again for these amazing improvements! :tada: :bow:

Could you rebase against master then we can merge this branch.

**Comment by ocourtin** [comment]
Daniel, 

Could you give a look on the rebase result.
My first with manuals conflicts to deal with...

Note for myself: create dedicated branches rather than keep working in my own master...

**Comment by daniel-j-h** [comment]
Looks good; flake8 and black are also fine. Going to merge this in.
